File: LeetCode/Study Plan/Practice Problems/Medium/Decode String/Go/decode_string.go
package main

import (
	"fmt"
	"strconv"
)

func decode_string(s string) string {
	str := []byte(s)
	n := len(str)
	stack := []byte{}
	for i := 0; i < n; {
		if str[i] == ']' {
			temp := ""
			j := len(stack) - 1
			for stack[j] != '[' {
				temp = string(stack[j]) + temp
				stack = stack[:j]
				j--
			}
			stack = stack[:j]
			j--
			num := ""
			for len(stack) > 0 && (stack[j] >= '0' && stack[j] <= '9') {
				num = string(stack[j]) + num
				stack = stack[:j]
				j--
			}
			val, _ := strconv.Atoi(num)
			t := ""
			for i := 0; i < val; i++ {
				t += temp
			}
			stack = append(stack, []byte(t)...)
		} else {
			stack = append(stack, str[i])
		}
		i++
	}
	return string(stack)
}

func main() {
	fmt.Println(decode_string("3[a]2[bc]"))
}

File: LeetCode/Medium/Decode String/Java/decodeString.java
class Solution {
    public String decodeString(String s) {
        Stack<Character> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();
        for (int i=0 ; i<s.length(); i++) {
            if (s.charAt(i) == ']') {
                StringBuilder temp = new StringBuilder();
                while (!stack.isEmpty() && stack.peek() != '[') temp.append(stack.pop());
                temp.reverse();
                stack.pop(); // to pop [
                String val = "";
                while (!stack.isEmpty() && Character.isDigit(stack.peek())) val = stack.pop() + val;
                
                if (val != "") {
                    Integer num = Integer.parseInt(val);
                    String t = temp.toString();
                    for (int k=0; k<num-1; k++) temp.append(t);
                }
                for (int k=0; k<temp.length(); k++) stack.push(temp.charAt(k));
            } else {
                stack.push(s.charAt(i));
            }
        }
        while(!stack.isEmpty()) sb.append(stack.pop());
        sb.reverse();
        return sb.toString();
    }
}
